tensorflow/c/c_api_experimental.cc

```diff
@@ -750,3 +756,325 @@ TF_Library* TF_LoadPluggableDeviceLibrary(const char* library_filename,
+                               void (*copyFunc)(TF_OpKernelContext * ctx, TF_Tensor *source, TF_Tensor *dest)) {
+
+  using namespace tensorflow;
+  auto* context = reinterpret_cast<::tensorflow::OpKernelContext*>(ctx);
+  if (copy_on_read_mode || !tensor->RefCountIsOne()) {
+    // Tensor's buffer is in use by some read, so we need to copy before
+    // updating.
+    Tensor tmp;
+    if(variantType) {
+      AllocatorAttributes attr;
+      attr.set_on_host(true);
+      TF_RETURN_IF_ERROR(
+          context->allocate_temp(tensor->dtype(), tensor->shape(), &tmp, attr));
+
+      const auto elements_in = tensor->flat<Variant>();
+      auto elements_out = tmp.flat<Variant>();
+      for (int64 i = 0; i < elements_in.size(); ++i) {
+        elements_out(i) = elements_in(i);
+      }
+    } else {
+      AllocatorAttributes attr;
+      attr.set_gpu_compatible(true);
+      attr.set_nic_compatible(true);
+      TF_RETURN_IF_ERROR(
+          context->allocate_temp(tensor->dtype(), tensor->shape(), &tmp, attr));
+      tensorflow::Status s;
+      TF_Tensor *tf_tmp = TF_TensorFromTensor(tmp, &s);
+      TF_Tensor *tf_tensor = TF_TensorFromTensor(*tensor, &s);
+      copyFunc(ctx, tf_tensor, tf_tmp);
+    }
+    *tensor = tmp;
+  }
+  return Status::OK();
+}
+
+tensorflow::mutex* GetTrainingVariableMutex(TF_OpKernelContext* ctx,
+                                            int32_t input,
+                                            bool sparse,
+                                  void (*copyFunc)(TF_OpKernelContext * ctx,
+                                                   TF_Tensor *source,
+                                                   TF_Tensor *dest),
+                                            tensorflow::Var** maybe_resource) {
+  auto* cc_ctx = reinterpret_cast<::tensorflow::OpKernelContext*>(ctx);
+  *maybe_resource = nullptr;
+  if (cc_ctx->input_dtype(input) == tensorflow::DT_RESOURCE) {
+    if (LookupResource(cc_ctx, HandleFromInput(cc_ctx, input), maybe_resource).ok()) {
+      if (sparse) {
+        EnsureSparseVariableAccess(ctx, false, copyFunc, *maybe_resource);
+      }
+      return (*maybe_resource)->mu();
+    } else {
+      cc_ctx->CtxFailureWithWarning(
+          tensorflow::errors::Internal("Invalid variable reference."));
+      return nullptr;
+    }
+  }
+  return cc_ctx->input_ref_mutex(input);
+}
+
+void TF_AssignVariable(TF_OpKernelContext* ctx,
+                       int input_index,
+                       int value_index,
+                       void (*copyFunc)(TF_OpKernelContext * ctx, TF_Tensor *source, TF_Tensor *dest),
+                       TF_Status* status) {
+  auto* cc_ctx = reinterpret_cast<::tensorflow::OpKernelContext*>(ctx);
+  tensorflow::AllocatorAttributes allocator_attr;
+  tensorflow::core::RefCountPtr<tensorflow::Var> variable;
+  const tensorflow::Tensor& value = cc_ctx->input(value_index);
+  OP_REQUIRES_OK(cc_ctx, LookupOrCreateResource<tensorflow::Var>(
```

Comment:
Technically this should be `tensorflow::LookupOrCreateResource`, but some of the headers might have defined `using tensorflow::LookupOrCreateResource` somewhere so the compiler didn't complain.

Reply:
Makes sense.